### ISSUE TITLE
Negative sampling for cold-source split

### ISSUE THREAD

**Issue by xiaobingbuhuitou** [carpenter-singh-lab/motive#6]
Hello, In the paper, it is mentioned that for the cold-source split, "the head of the negative edge is sampled from the unique sources in the supervision ST edges in the batch, and the tail of the negative edge is sampled from ST edges in the batch." Combined with the algorithm described in the paper, does the negative sampling for the cold-source split in the train/valid/test phases involve selecting source nodes with only one edge in the current batch and combining them with target nodes from the entire dataset to form negative edges, while removing those that already exist in the ST edges? Could you please tell me what "supervision ST edges in the batch" means?

**Comment by johnarevalo** [comment]
Thanks for asking this. You are right, except that:

 - _selecting source nodes with only one edge_ : No. the node degree is not considered in the negative sampling algorithm. However, the probability of selecting a source node as the head of a negative edge is implicitly influenced by the mini-batch sampling from PyG.
 -  _combining them with target nodes from the entire dataset to form negative edges_: Not from the entire dataset, but from the target nodes present in the current mini-batch.

"supervision ST edges in the batch", refers to the set of source-target edges that are used in the loss function to compute the prediction error. Recall that in GNN the set of training edges should be split in "message edges" and "supervision edges", first ones allow message passing, while second ones are used for error computation. You can check more details at [Setting-up GNN prediction tasks](https://web.stanford.edu/class/cs224w/slides/05-GNN3.pdf) Slide-46
